File: src/test/js/Validator.test.js
const R = require('ramda')
const _ = require('lodash')
const {User} = require("../../main/js/User")
const {Validator, pipe, compose} = require("../../main/js/Validator")
const Gender = require("../../main/js/Gender")

describe('Validator Test', function () {

    test("validation", function () {

        let user = new User("", 90, null, "foobar.com")

        const result = Validator.of(user)
            .mapAndValidate(user => user._name, name => name == null, "name is null")
            .mapAndValidate(user => user._name, name => name == null || name.length === 0, "name is empty")
            .mapAndValidate(user => user._age, age => !(age > 20 && age < 30), "age isn't between...")
            .mapAndValidate(user => user._gender, gender => gender == null, "gender is null")
            .mapAndValidate(user => user._email, email => !email.includes("@"), "email doesn't contains '@'")
            .get()

        result.forEach(it => console.log(it))
    })

    test("fun", function () {
        let user = new User("John", 24, Gender.MALE, "foobar.com")
        let name = user => user._name
        console.log(name(user))
    })

    test("pipe", function () {
        let user = new User("John", 24, Gender.MALE, "foobar.com")
        let name = user => user._name;
        let czyEmpty = s => s != null && !s.isEmpty
        console.log(pipe(name, czyEmpty)(user))
    })

    test("compose", function () {
        let user = new User("John", 24, Gender.MALE, "foobar.com")
        let name = user => user._name;
        let czyEmpty = s => s != null && !s.isEmpty
        console.log(compose(czyEmpty, name)(user))
    })

})

File: src/test/java/org/example/ValidatorTest.java
package org.example;

import java.util.Objects;
import java.util.function.Predicate;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ValidatorTest {

    @Test
    void givenInvalidUser_whenValidatorValidate_thenReturnErrors() {
        User user = new User("", 90, null, "foobar.com");
        var result = Validator.of(user)
                .validate(User::name, Objects::isNull, "name is null")
                .validate(User::name, String::isEmpty, "name is empty")
                .validate(User::gender, Objects::isNull, "gender is null")
                .validate(User::email, email -> !email.contains("@"), "email doesn't contains '@'")
                .validate(User::age, age -> !(age > 20 && age < 30), "age isn't between...")
                .get();
        result.forEach(System.out::println);
        Assertions.assertFalse(result.isEmpty());
    }

    @Test
    void givenValidUser_whenValidate_thenReturnNoErrors() {
        User user = new User("", 90, null, "foobar.com");
        var result = Validator.of(user)
                .validate(Objects::isNull, "user is null")
                .validate(user1 -> user1.email() == null, "email is null")
                .get();
        result.forEach(System.out::println);
        Assertions.assertTrue(result.isEmpty());
    }

    @Test
    void givenInvalidUser_whenPredicateTest_thenReturnFalse() {
        User user = new User("", 90, null, "foobar.com");
        Predicate<User> predicate = Validator.validate((User u) -> u.age(), age -> age > 20 && age < 30);
        var result = predicate.test(user);
        Assertions.assertFalse(result);
    }

    @Test
    void givenValidUser_whenPredicateTest_thenReturnTrue() {
        User user = new User("", 90, null, "foobar.com");
        Predicate<User> predicate = Validator.isEqual((User u) -> u.age(), 90);
        var result = predicate.test(user);
        Assertions.assertTrue(result);
    }

}